language-plutus-core/src/Language/PlutusCore/Renamer.hs

```diff
@@ -111,104 +112,79 @@ annotateTy (TyApp x ty ty') =
-renameType _ ty@TyInt{} = pure ty
-renameType st ty@(TyLam x (TyName (Name x' s (Unique u))) k ty') = do
-    m <- get
-    let st' = modifyIdentifiers u m st
-        pastDef = lookupId u st
-    case pastDef of
-        Just _ ->
-            modify (+1) >>
-            TyLam x (TyName (Name x' s (Unique (m+1)))) k <$> renameType st' ty'
-        _ -> renameType st' ty
-renameType st ty@(TyForall x (TyName (Name x' s (Unique u))) k ty') = do
-    m <- get
-    let st' = modifyIdentifiers u m st
-        pastDef = lookupId u st
-    case pastDef of
-        Just _ ->
-            modify (+1) >>
-            TyForall x (TyName (Name x' s (Unique (m+1)))) k <$> renameType st' ty'
-        _ -> renameType st' ty
-renameType st ty@(TyFix x (TyName (Name x' s (Unique u))) ty') = do
-    m <- get
-    let st' = modifyIdentifiers u m st
-        pastDef = lookupId u st
-    case pastDef of
-        Just _ ->
-            modify (+1) >>
-            TyFix x (TyName (Name x' s (Unique (m+1)))) <$> renameType st' ty'
-        _ -> renameType st' ty
-renameType st ty@(TyVar x (TyName (Name x' s (Unique u)))) =
-    case pastDef of
-        Just j -> pure $ TyVar x (TyName (Name x' s (Unique j)))
-        _      -> pure ty
-    where pastDef = lookupId u st
-renameType st (TyApp x ty ty') = TyApp x <$> renameType st ty <*> renameType st ty'
-renameType st (TyFun x ty ty') = TyFun x <$> renameType st ty <*> renameType st ty'
-renameType _ ty@TyBuiltin{} = pure ty
+-- | Mapping from locally unique indices to globally unique 'Unique's.
+newtype UniquesRenaming = UniquesRenaming
+    { unUniquesRenaming :: IM.IntMap Unique
+    }
+
+-- | The monad the renamer runs in.
+type RenameM = ReaderT UniquesRenaming Quote
+
+-- | Rename 'Unique's so that they're globally unique.
+rename :: (Rename a, MonadQuote m) => a -> m a
+rename x = liftQuote $ runReaderT (renameM x) (UniquesRenaming mempty)
+
+-- | Save the mapping from an old 'Unique' to a new one.
+updateUniquesRenaming :: Unique -> Unique -> UniquesRenaming -> UniquesRenaming
+updateUniquesRenaming uniqOld uniqNew (UniquesRenaming uniques) =
+    UniquesRenaming $ IM.insert (unUnique uniqOld) uniqNew uniques
+
+-- | Look up a new 'Unique' an old 'Unique' got mapped to.
+lookupUnique :: Unique -> RenameM (Maybe Unique)
+lookupUnique uniq = asks $ IM.lookup (unUnique uniq) . unUniquesRenaming
+
+-- | Replace the 'Unique' in a value by a new 'Unique', save the mapping
+-- from an old 'Unique' to the new one and supply the updated value to a continuation.
+withRefreshed :: HasUnique a => a -> (a -> RenameM c) -> RenameM c
+withRefreshed x k = do
+    let uniqOld = x ^. unique
+    uniqNew <- liftQuote freshUnique
+    local (updateUniquesRenaming uniqOld uniqNew) $ k (x & unique .~ uniqNew)
+
+-- | The class of things that can be renamed.
+-- I.e. things that are capable of satisfying the global uniqueness condition.
+class Rename a where
+    renameM :: a -> RenameM a
```

Comment:
I'm not a fan of this typeclass. What's the advantage of this over having `renameTerm` and `renameType`?

Reply:
The renamer is now written in the same style as the cloning machinery. There are no big advantages I see, but small are:

1. since names are abstract, it's more clear what's happening, because we deal with `Unique`s rather than particular names
2. in case names change (and I think I would like them to change, because I do not understand why we have `annotateTerm` instead of keeping track of variables in an environment in the type checker and this is the only place where we need to distinguish type variable names from term variable names, right?), we won't need to change most of what is written there
3. we do not currently rename anything that doesn't have names other than `TyName` and `Name`, but in case we want to do this in future, we won't have any problems

Reply:
Yeah, I think the ability to be generic over different kinds of names is nice. It'll also make it easier to be generic over renaming dynamic builtins.

Reply:
Right, but the cloning machinery needed the added genericity. 

My main complaint is that this makes things slightly harder to follow with no advantage as fr as I can tell.

As for 1., I think we can abstract among names using `HasUnique` and `RenameM` alone without need for `class Rename a`, unless I am missing something. 

> I do not understand why we have annotateTerm instead of keeping track of variables in an environment in the type checker

I believe this is the standard approach in compilers and it was suggested by Manuel. 

@michaelpj we can be generic over names using `HasUnique` alone, no?

Reply:
> My main complaint is that this makes things slightly harder to follow with no advantage as fr as I can tell.

What is harder about a simple Haskell 98 class? We have a single `rename` function that is used for renaming `Program`, `Term` and `Type` -- we need to rename all those. I do not want to think "ok, this is a `Type`, so I'm gonna use `renameType`, and this is a `Term`, so `renameTerm` is what I need". I just write `rename` uniformly in the stdlib and everything works. And I write it for each definition there (well, not at the moment, but that's the plan). So I highly prefer to have this class.

> As for 1., I think we can abstract among names using HasUnique and RenameM alone without need for class Rename a, unless I am missing something.

This is true, I originally wrote it like that. Looking at it again, I think using just `HasUnique` is indeed a better approach. I'll prepare another PR.

Reply:
I think the broader style question is "is it okay to use a typeclass for ad-hoc polymorphism" to which I think the answer is "judiciously". This seems fine to me, but I do appreciate the perspective that typeclasses should be only for stuff with laws etc.

Note that we will want to abstract over dynamic builtins in the future, which I think will need `Rename`.